sedaro/src/sedaro/results/utils.py

```diff
@@ -260,3 +320,100 @@ def from_file(self, filename: Union[str, Path]):
+
+    @staticmethod
+    def agent_name_from_filename(filename: str):
+        '''Convert a file name back to the original agent name.'''
+        return filename.replace('.', '/')
+
+    def save_parquets(self, data: dict, path: Union[str, Path]):
+        '''Save the DataFrames as Parquet files in their proper location given the specified save path.'''
+        if TYPE_CHECKING:
+            import dask.dataframe as dd
+        os.mkdir(data_subdir_path := self.data_subdir(path))
+        parquet_files = []
+        for agent in data:
+            agent_parquet_path = f"{data_subdir_path}/{(pq_filename := self.agent_name_for_filename(agent))}"
+            parquet_files.append(pq_filename)
+            df: dd = data[agent]
+            try:
+                df.to_parquet(agent_parquet_path)
+            except ValueError as e:
+                # the ValueError contains details on the schema mismatch, in a predictable format
+                # REF: https://github.com/dask/dask/blob/main/dask/dataframe/io/parquet/arrow.py#L786-L798
+                if str(e).startswith("Failed to convert partition to expected pyarrow schema:"):
+                    harmonized_df_schema = df_schema_for_pyarrow(e)
+                    df.to_parquet(agent_parquet_path, schema=harmonized_df_schema)
+                else:
+                    raise e
+        return parquet_files
+
+    @classmethod
+    def load_parquets(cls, path: Union[str, Path], metadata: dict) -> dict:
+        '''Load Parquet files from the specified save path and return a dictionary of DataFrames.'''
+        import dask.dataframe as dd
+        data_subdir_path = cls.data_subdir(path)
+        data = {}
+        try:
+            parquet_files = metadata['parquet_files']
+        except KeyError:
+            parquet_files = get_parquets(data_subdir_path)
+        for agent in parquet_files:
+            df = dd.read_parquet(f"{data_subdir_path}/{agent}")
+            data[cls.agent_name_from_filename(agent)] = df
+        return data
+
+    def save(self, path: Union[str, Path]):
+        '''Save the {class_name}'s data to a directory with the specified path.'''
+        try:
+            os.makedirs(path)
+        except FileExistsError:
+            if not (os.path.isdir(path) and any(os.scandir(path))):
+                raise FileExistsError(
+                    f"A file or non-empty directory already exists at {path}. Please specify a different path.")
+        with open(f"{path}/class.json", "w") as fp:
+            json.dump({'class': self.__class__.__name__}, fp)
+        metadata_to_save = self.do_save(path)
+        with open(f"{path}/meta.json", "w") as fp:
+            json.dump(metadata_to_save, fp)
+        print(f"{self.__class__.__name__} saved to {path}.")
+
+    @abstractmethod
+    def do_save(self, path: Union[str, Path]) -> dict:
+        '''Save the data to the specified path. Return a metadata dict to be saved alongside it.'''
+        pass
```

Comment:
Are the `do_save` and `do_load` methods meant to be called by the end users? If not, what would you think about indicating that these methods are "internal" with a [single underscore](https://peps.python.org/pep-0008/#:~:text=_single_leading_underscore) (such as `_do_save` and `_do_load`)? This can help the users more quickly find useful methods in the intellisense

I don't know enough to say, but if there are only new methods here only meant for internal use, perhaps those could be updated as well.